File: sdk/python/rabbittwatch/integrations/langchain.py
import json

try:
    from langchain_core.tracers.base import BaseTracer
    from langchain_core.tracers.schemas import Run
    HAS_LANGCHAIN = True
except ImportError:
    class BaseTracer:
        pass
    class Run:
        pass
    HAS_LANGCHAIN = False

class RabbittWatchLangChainTracer(BaseTracer):
    def __init__(self, rw_ai, environment="prod", service_name="app", tags=None):
        if not HAS_LANGCHAIN:
            raise ImportError("langchain-core is required to use RabbittWatchLangChainTracer. Install it using pip install langchain-core")
        super().__init__()
        self.rw_ai = rw_ai
        self.environment = environment
        self.service_name = service_name
        self.tags = tags or []

    def _persist_run(self, run: Run) -> None:
        spans = []
        self._collect_spans(run, None, spans)
        
        started_at = run.start_time.isoformat() + "Z" if run.start_time else None
        ended_at = run.end_time.isoformat() + "Z" if run.end_time else None
        
        envelope = {
            "trace": {
                "traceId": str(run.id),
                "name": run.name,
                "serviceName": self.service_name,
                "environment": self.environment,
                "startedAt": started_at,
                "endedAt": ended_at,
                "tags": self.tags,
            },
            "spans": spans,
        }
        
        self.rw_ai.client.enqueue(envelope)
        self.rw_ai.client.flush()

    def _collect_spans(self, run: Run, parent_span_id, spans: list):
        started_at = run.start_time.isoformat() + "Z" if run.start_time else None
        ended_at = run.end_time.isoformat() + "Z" if run.end_time else None
        duration_ms = int((run.end_time - run.start_time).total_seconds() * 1000) if (run.end_time and run.start_time) else 0
        
        kind = "custom"
        if run.run_type == "llm":
            kind = "llm"
        elif run.run_type == "chain":
            kind = "chain"
        elif run.run_type == "tool":
            kind = "tool"
            
        prompt_tokens, completion_tokens, total_tokens = self._get_tokens(run)
        
        capture_input = self.rw_ai.options.get("captureInput", False)
        capture_output = self.rw_ai.options.get("captureOutput", False)
        max_chars = self.rw_ai.options.get("previewMaxChars", 102400)
        
        input_preview = None
        if capture_input and run.inputs:
            input_preview = json.dumps(run.inputs)[:max_chars]
            
        output_preview = None
        if capture_output and run.outputs:
            output_preview = json.dumps(run.outputs)[:max_chars]
            
        provider = None
        model = None
        if run.extra:
            metadata = run.extra.get("metadata") or {}
            invocation_params = run.extra.get("invocation_params") or {}
            provider = metadata.get("ls_provider") or invocation_params.get("_type")
            model = metadata.get("ls_model_name") or invocation_params.get("model_name") or invocation_params.get("model")

        spans.append({
            "spanId": str(run.id),
            "parentSpanId": str(parent_span_id) if parent_span_id else None,
            "name": run.name,
            "kind": kind,
            "provider": provider,
            "model": model,
            "status": "error" if run.error else "success",
            "startedAt": started_at,
            "endedAt": ended_at,
            "durationMs": duration_ms,
            "promptTokens": prompt_tokens,
            "completionTokens": completion_tokens,
            "totalTokens": total_tokens,
            "errorMessage": run.error,
            "inputPreview": input_preview,
            "outputPreview": output_preview,
            "metadata": {
                "run_type": run.run_type,
                "extra": run.extra,
            }
        })
        
        if run.child_runs:
            for child in run.child_runs:
                self._collect_spans(child, run.id, spans)

    def _get_tokens(self, run: Run):
        prompt_tokens = 0
        completion_tokens = 0
        if run.run_type == "llm" and run.outputs:
            generations = run.outputs.get("generations")
            token_usage = None
            if generations and len(generations) > 0 and len(generations[0]) > 0:
                gen_info = generations[0][0].get("generation_info") or {}
                token_usage = gen_info.get("token_usage") or gen_info.get("usage_metadata")
            
            if not token_usage and run.outputs.get("llm_output"):
                token_usage = run.outputs.get("llm_output", {}).get("token_usage")
                
            if token_usage:
                prompt_tokens = token_usage.get("prompt_tokens") or token_usage.get("input_tokens") or token_usage.get("promptTokens") or 0
                completion_tokens = token_usage.get("completion_tokens") or token_usage.get("output_tokens") or token_usage.get("completionTokens") or 0
                
        return prompt_tokens, completion_tokens, prompt_tokens + completion_tokens

File: sdk/js/src/integrations/langchain.ts
import { BaseTracer, Run } from "@langchain/core/tracers/base";
import { RabbittWatchAI } from "../index.js";
import { TraceEnvelope } from "../types.js";

export class RabbittWatchLangChainTracer extends BaseTracer {
    name = "RabbittWatchLangChainTracer";

    constructor(
        private rwAI: RabbittWatchAI,
        private options: {
            environment?: string;
            serviceName?: string;
            tags?: string[];
        } = {}
    ) {
        super();
    }

    protected async persistRun(run: Run): Promise<void> {
        const spans: any[] = [];
        this.collectSpans(run, null, spans);

        const startedAt = new Date(run.start_time).toISOString();
        const endedAt = run.end_time ? new Date(run.end_time).toISOString() : new Date().toISOString();

        const envelope: TraceEnvelope = {
            trace: {
                traceId: run.id,
                name: run.name,
                serviceName: this.options.serviceName || (this.rwAI as any).options.serviceName || "app",
                environment: this.options.environment || (this.rwAI as any).options.environment || "prod",
                startedAt,
                endedAt,
                tags: this.options.tags,
            },
            spans,
        };

        // Enqueue trace data to TelemetryClient
        (this.rwAI as any).client.enqueue(envelope);
    }

    private collectSpans(run: Run, parentSpanId: string | null, spans: any[]) {
        const startedAt = new Date(run.start_time).toISOString();
        const endedAt = run.end_time ? new Date(run.end_time).toISOString() : undefined;
        const durationMs = run.end_time ? run.end_time - run.start_time : 0;

        let kind: "chain" | "tool" | "llm" | "custom" = "custom";
        if (run.run_type === "llm") kind = "llm";
        else if (run.run_type === "chain") kind = "chain";
        else if (run.run_type === "tool") kind = "tool";

        const { promptTokens, completionTokens, totalTokens } = this.getTokens(run);

        const captureInput = (this.rwAI as any).options.captureInput;
        const captureOutput = (this.rwAI as any).options.captureOutput;
        const maxChars = (this.rwAI as any).options.previewMaxChars ?? 102400;

        const inputPreview = captureInput && run.inputs ? JSON.stringify(run.inputs).slice(0, maxChars) : undefined;
        const outputPreview = captureOutput && run.outputs ? JSON.stringify(run.outputs).slice(0, maxChars) : undefined;

        spans.push({
            spanId: run.id,
            parentSpanId,
            name: run.name,
            kind,
            provider: run.extra?.metadata?.ls_provider || run.extra?.invocation_params?._type || undefined,
            model: run.extra?.metadata?.ls_model_name || run.extra?.invocation_params?.model_name || undefined,
            status: run.error ? "error" : "success",
            startedAt,
            endedAt,
            durationMs,
            promptTokens,
            completionTokens,
            totalTokens,
            errorMessage: run.error || undefined,
            inputPreview,
            outputPreview,
            metadata: {
                run_type: run.run_type,
                extra: run.extra,
            }
        });

        if (run.child_runs) {
            for (const child of run.child_runs) {
                this.collectSpans(child, run.id, spans);
            }
        }
    }

    private getTokens(run: Run) {
        let promptTokens = 0;
        let completionTokens = 0;

        if (run.run_type === "llm") {
            const usage = run.outputs?.llmOutput?.tokenUsage || 
                          run.outputs?.tokenUsage ||
                          run.outputs?.generations?.[0]?.[0]?.generationInfo?.tokenUsage ||
                          run.outputs?.generations?.[0]?.[0]?.message?.usage_metadata;
            if (usage) {
                promptTokens = usage.promptTokens || usage.prompt_tokens || usage.input_tokens || 0;
                completionTokens = usage.completionTokens || usage.completion_tokens || usage.output_tokens || 0;
            }
        }

        return {
            promptTokens,
            completionTokens,
            totalTokens: promptTokens + completionTokens
        };
    }
}
